synapse/lib/ast.py

```diff
@@ -536,19 +549,21 @@ async def todogenr():
 
                 if doedges:
 
-                    async for buid01 in results:
-
+                    async for n2buid, verb in runt.snap.iterEdgePref(node.buid):
                         await asyncio.sleep(0)
+                        if rev.get(n2buid) is None:
+                            rev[n2buid] = collections.defaultdict(list)
 
-                        iden01 = s_common.ehex(buid01)
-                        async for verb in node.iterEdgeVerbs(buid01):
-                            await asyncio.sleep(0)
-                            edges.append((iden01, {'type': 'edge', 'verb': verb}))
+                        # TODO: old way doesn't make self-referential nodes?
+                        if n2buid != node.buid:
+                            rev[n2buid][node.buid].append(verb)
+                            if n2buid in results:
+                                edges.append((s_common.ehex(n2buid), {'type': 'edge', 'verb': verb}))
 
-                        # for existing nodes, we need to add n2 -> n1 edges in reverse
-                        async for verb in runt.snap.iterEdgeVerbs(buid01, node.buid):
```

Comment:
Not using `iterEdgeVerbs()` seems like we might be back to doing brute force or caching a number of things in RAM that could be prohibitive?

Reply:
So we do end up caching a fair bit. That's because the longest pole in the tent was going down into the slab and constructing the `Scan` object and hitting both it's constructor and `__enter__` methods when we do `iterEdgeVerbs`. We actually end up hitting `iterEdgeVerbs`  around O(n^2) times, which is super costly when usually we end up returning nothing. Nothing else in the function is even close to as costly as that, so just caching things and cutting those calls is the main reason the graph projection in this PR is a lot faster.

We might be able to throw things into a spooled set to help with the memory pressure, but that was something I wanted to discuss before throwing it into this PR since I wanted this PR to be relatively fast and easy.